File: learning_topic/src/pose_subscriber.cpp
/**
 * 该例程将订阅 /turtle1/pose 话题，消息类型 turtlesim::Pose
 */

#include <ros/ros.h>
#include "turtlesim/Pose.h"

// 接收到订阅的消息后，会进入消息回调函数
void poseCallback(const turtlesim::Pose::ConstPtr& msg)
{
    // 将接收到的消息打印出来
    ROS_INFO("Turtle pose: x:%0.6f, y:%0.6f", msg->x, msg->y);
}

int main(int argc, char **argv)
{
    // 初始化 ROS 节点
    ros::init(argc, argv, "pose_subscriber");

    // 创建节点句柄
    ros::NodeHandle n;

    // 创建一个 Subscriber，订阅名为 /turtle1/pose 的 topic， 注册回调函数 poseCallback
    ros::Subscriber pose_sub = n.subscribe("/turtle1/pose", 10, poseCallback);

    // 循环等待回调函数
    ros::spin();

    return 0;
}

File: learning_topic/scripts/pose_subscriber.py
# ! /usr/bin/env python3
# -*- coding: utf-8 -*-
# 该例程将订阅 /turtle1/pose 话题，消息类型 turtlesim::Pose

import rospy
from turtlesim.msg import Pose

def poseCallback(msg):
    rospy.loginfo("Turtle pose: x:%0.6f, y%0.6f", msg.x, msg.y)

def pose_subscriber():
    # ROS 节点初始化
    rospy.init_node("pose_subscriber", anonymous=True)

    # 创建一个 Subscriber，订阅名为 /turtle1/pose 的 topic，注册回调函数 poseCallback
    rospy.Subscriber("/turtle1/pose", Pose, poseCallback)

    # 循环等待回调函数
    rospy.spin()

if __name__ == "__main__":
    pose_subscriber()
    